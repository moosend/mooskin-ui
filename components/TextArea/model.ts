import { ITextAreaBoxComponentProps } from '../Box/model';

export interface ITextAreaComponentProps extends ITextAreaBoxComponentProps {
	/** what data is being used, helps whn extracting user input, you know on what field changes are made */
	dataLabel?: string;
}
