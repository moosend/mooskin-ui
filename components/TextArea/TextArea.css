@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.areaContainer{
    display: flex;
    flex-grow: 1;
    flex-shrink: 1;
}

.textAreaDiv{
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    flex-shrink: 1;
    box-sizing: border-box;
}

.textarea{
    outline: none;
    padding: 7px;
    font-family: Arial,sans-serif;
    border: 0;
    color: var(--inputFontColor);
    font-size: 1rem;
    border-radius: 8px;
    background: #efefef;
    width: 100%;
    height:100%;
    transition: 0.2s;
}

.textarea::-webkit-scrollbar {
    width: 5px;
    height: 10px;
}

.textarea::-webkit-scrollbar-button:start:decrement,
.textarea::-webkit-scrollbar-button:end:increment  {
    display: none;
}

.textarea::-webkit-scrollbar-track-piece  {
    background-color: #eeeeee;
}

textarea::-webkit-scrollbar-thumb:vertical {
    background: #d9d9d9;
    border-radius: 16px;
}

.disabledTextArea{
    cursor:not-allowed;
    color: #bfbfbf;
    background: #f6f6f6;
    resize: none;
}

.textarea::placeholder {
    color: var(--inputPlaceholder);
}

.textarea:focus {
    outline: none;
    color:#333333;
}

.textareaLabel {
    color: var(--inputLabel);
    /* font-size: 13px;*/
    font-size: .92857142857rem;
    padding-top: 7px;
    cursor: pointer;
    flex-basis: 150px;
}

.description{
    display: block;
    color: #b3b3b3;
    font-size: .78571428571rem;
    margin-top: 10px;
}

.success{
    box-shadow: inset 0px 0px 0px 2px #00cc44;
}

.error{
    box-shadow: inset 0px 0px 0px 2px #ec3232;
}

.descError{
    color: #ec3232;
}

.descSuccess{
    color: #00cc44;
}

@media (--xSmall){
    .areaContainer{
        flex-direction: column
    }
    
    .textareaLabel{
        flex-basis: 10px;
        padding-top: 0px;
        margin-left: 3px;
    }

    :global(.mooskin-textarea-editor .rdw-colorpicker-modal){
        position: fixed;
        top: 25%;
        left: 15%;
    }

    :global(.mooskin-textarea-editor .rdw-link-modal){
        position: fixed;
        top: 25%;
        left: 20%;
    }

    :global(.mooskin-textarea-editor .rdw-embedded-modal){
        position: fixed;
        top: 25%;
        left: 20%;
    }

    :global(.mooskin-textarea-editor .rdw-emoji-modal){
        position: fixed;
        top: 25%;
        left: 20%;
    }

    :global(.mooskin-textarea-editor .rdw-image-modal){
        position: fixed;
        top: 25%;
        left: 20%;
    }

}


@media (--deviceXsmall){
    .toolbar{
        position: fixed;
        top: 0;
        left: 0;
        right: 0;
    }
}
