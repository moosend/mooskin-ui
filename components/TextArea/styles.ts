import styled from 'styled-components';

// Models
import { ITextAreaBoxComponentProps } from '../Box/model';

// Components
import { Box } from '../Box/Box';

// "CSS" variables
import variables from '../_utils/globals/variables';

export const StyledTextArea = styled(Box)<ITextAreaBoxComponentProps>`
	padding: 11px;
	border-radius: 2px;
	color: ${(props) => props.palette?.fontColors.text || variables.fontColors.text};
	border: solid 1px ${(props) => props.palette?.borderColors.gray1 || variables.borderColors.gray1};
	background-color: ${(props) => props.palette?.backgroundColors.white || variables.backgroundColors.white};
	opacity: ${(props) => (props.disabled ? 0.5 : 1)};
	font-size: 12px;
`;

StyledTextArea.displayName = 'StyledTextArea';
