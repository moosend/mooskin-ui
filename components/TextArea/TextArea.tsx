import * as React from 'react';

import styles from './TextArea.css';

import {IInputCallbackData, IValidationCallbackData} from '../_utils/types/commonTypes';

export interface ITextAreaProps {

    /** override textarea id */
    id?: string;

    /** provide to make the textarea field disabled */
    disabled?: boolean;

    /** provide to make the textarea field required */
    required?: boolean;

    /** specify textarea columns */
    cols?: number;

    /** specify textarea rows */
    rows?: number;

    /** override textarea name */
    name?: string;

    /** override textarea value */
    value?: string;

    /** override textarea placeholder */
    placeholder?: string;

    /** override textarea minlength */
    minlength?: number;

    /** override textarea maxlength */
    maxlength?: number;

    /** textarea label */
    label?: string;

    /** textarea description (small italic bottom) */
    description?: string;

    /** spacing between label and textarea */
    labelWidth?: number;

    /** toggle readonly textarea */
    readonly?: boolean;

    /** override textarea styles */
    style?: React.CSSProperties;

    /** override textarea class */
    className?: string;

    /** status of the textarea, error or success */
    status?: 'error' | 'success' | undefined;

    /** validate function */
    validate?: (data: IValidationCallbackData) => boolean;

    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** callback that is called when the textarea changes */
    onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>, data: IInputCallbackData) => void;

}

class TextArea extends React.Component<ITextAreaProps, {}> {

    static defaultProps = {
        className: '',
        style: {}
    };

    id: string;

    constructor(props: ITextAreaProps){
        super(props);

        this.id = this.props.id || this.generateId();

    }

    render(){

        const disabledtextarea = this.props.disabled ? styles.disabledTextArea : '';
        const status = this.getStatus();

        return (
            <div
                className={`textarea-component ${styles.areaContainer} ${this.props.className}`}
                style={this.props.style}
            >
                {this.props.label && this.getLabel()}
                <div className={styles.textAreaDiv}>
                    <textarea
                        id={this.id}
                        onChange={this.onChange}
                        name={name}
                        value={this.props.value}
                        placeholder={this.props.placeholder}
                        cols={this.props.cols}
                        rows={this.props.rows}
                        minLength={this.props.minlength}
                        maxLength={this.props.maxlength}
                        required={this.props.required}
                        disabled={this.props.disabled}
                        readOnly={this.props.readonly}
                        className={`textarea ${styles.textarea} ${disabledtextarea} ${status}`}
                        onBlur={this.validateOnBlur}
                    />
                    {this.props.description && this.getDescription()}
                </div>
            </div>
        );
    }

    getLabel = () => {
        const spacing = !this.props.labelWidth ? {} : {flexBasis: `${this.props.labelWidth}px`};
        return (
            <label
                className={styles.textareaLabel}
                style={spacing}
                htmlFor={this.id}
            >
                {this.props.label}
            </label>
        );
    }

    getDescription = () => {
        const descStatus = this.getDescStatus();
        return (
            <i
                className={`${styles.description} ${descStatus}`}
            >
                {this.props.description}
            </i>
        );
    }

    onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        !this.props.disabled &&
        this.props.onChange &&
        this.props.onChange(e, {value: e.target.value, dataLabel: this.props.dataLabel});
        if (this.props.status){
            this.props.validate &&
            this.props.validate({
                dataLabel: this.props.dataLabel,
                required: this.props.required,
                value: e.target.value
            });
        }
    }

    validateOnBlur = () => {
            this.props.validate &&
            this.props.validate({
                dataLabel: this.props.dataLabel,
                required: this.props.required,
                value: this.props.value
            });
    }

    getStatus = () => {
        const textStatus = this.props.status && this.props.status;
        if (textStatus){
            if (textStatus === 'error'){
                return styles.error;
            } else if (textStatus === 'success'){
                return styles.success;
            }
        }
    }

    getDescStatus = () => {
        const textStatus = this.props.status && this.props.status;
        if (textStatus){
            if (textStatus === 'error'){
                return styles.descError;
            } else if (textStatus === 'success'){
                return styles.descSuccess;
            }
        }
    }

    generateId = () => {
        return Math.random().toString(36).substr(2, 10);
    }
}

export default TextArea;
