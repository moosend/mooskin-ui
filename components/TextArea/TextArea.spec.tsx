import * as React from 'react';
import { TextArea } from './TextArea';

import { mount } from 'enzyme';

describe('TextArea', () => {
	test('renders correctly', () => {
		const func = jest.fn();

		const tree = mount(<TextArea value="asd" onChange={func} />);
		expect(tree).toMatchSnapshot();
	});

	test('renders properly into dom and has Placeholder prop', () => {
		const component = mount(<TextArea value="value" placeholder="username" />);

		expect(component.find('StyledTextArea').prop('placeholder')).toContain('username');
	});

	test('renders an input with a "required" prop and minlength', () => {
		const component = mount(<TextArea value="value" minLength={5} required />);

		expect(component.find('StyledTextArea').prop('required')).toEqual(true);
		expect(component.find('StyledTextArea').prop('minLength')).toEqual(5);
	});

	test('renders an input with a passed value and maxlength', () => {
		const component = mount(<TextArea value="random" maxLength={50} />);

		expect(component.find('StyledTextArea').prop('value')).toEqual('random');
		expect(component.find('StyledTextArea').prop('maxLength')).toEqual(50);
	});

	test('renders an input with custom css class and style', () => {
		const component = mount(<TextArea value="value" style={{ color: 'blue' }} className="input-group" />);

		expect(component.find('StyledTextArea').hasClass('input-group')).toBe(true);
		expect(component.find('StyledTextArea').prop('style')).toEqual({ color: 'blue' });
	});

	test('onChange prop callback is called when a key is pressed', () => {
		const func = jest.fn();

		const component = mount(<TextArea value="value" onChange={func} />);

		component.find('StyledTextArea').simulate('change', { target: { value: 'text' } });
		expect(func).toHaveBeenCalled();
	});
});
