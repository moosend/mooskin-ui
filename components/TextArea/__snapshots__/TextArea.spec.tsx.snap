// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
<TextArea
  className="myClass"
  cols={40}
  description="just some textarea"
  disabled={true}
  id="textarea"
  label="with a label ofc"
  labelWidth={30}
  maxlength={50}
  minlength={10}
  name="textarea name"
  onChange={[Function]}
  placeholder="textarea placeholder"
  required={true}
  rows={10}
  style={
    Object {
      "color": "blue",
    }
  }
  value="textarea value"
>
  <div
    className="textarea-component areaContainer myClass"
    style={
      Object {
        "color": "blue",
      }
    }
  >
    <label
      className="textareaLabel"
      htmlFor="textarea"
      style={
        Object {
          "flexBasis": "30px",
        }
      }
    >
      with a label ofc
    </label>
    <div
      className="textAreaDiv"
    >
      <textarea
        className="textarea textarea disabledTextArea undefined"
        cols={40}
        disabled={true}
        id="textarea"
        maxLength={50}
        minLength={10}
        name="nodejs"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="textarea placeholder"
        required={true}
        rows={10}
        value="textarea value"
      />
      <i
        className="description undefined"
      >
        just some textarea
      </i>
    </div>
  </div>
</TextArea>
`;
